File: old/experiments/latent_variable_experiment.py
from csv import reader
import numpy as np
from matplotlib import pyplot as plt

# prior = True
# num_runs = 1000

prior = False
num_samples = 200


N = 30
t = range(0,51,5)

to_plot = 1

postfix = '_posterior_final'
if prior:
	postfix = '_prior'


cum_vals = np.zeros((num_samples,2*N))
with open('latent_variable_cum_vals'+postfix+'.csv','rb') as csvfile:
	csvreader = reader(csvfile)
	for n, row in enumerate(csvreader):
		cum_vals[n,:] = np.array([int(v) for v in row])

times = np.zeros((num_samples,2*N))
with open('latent_variable_times'+postfix+'.csv','rb') as csvfile:
	csvreader = reader(csvfile)
	for n, row in enumerate(csvreader):
		times[n,:] = np.array([float(v) for v in row])

ys = np.zeros((num_samples,len(t)))
with open('latent_variable_ys'+postfix+'.csv','rb') as csvfile:
	csvreader = reader(csvfile)
	for n, row in enumerate(csvreader):
		ys[n,:] = np.array([float(v) for v in row])		

with open('latent_variable_mean_ps_ks'+postfix+'.csv','rb') as csvfile:
	csvreader = reader(csvfile)
	ps = np.array([float(v) for v in csvreader.next()])
	ks = np.array([int(v) for v in csvreader.next()])	

with open('latent_variable_mean_ps_zonn_ks'+postfix+'.csv','rb') as csvfile:
	csvreader = reader(csvfile)
	ps_zonn = np.array([float(v) for v in csvreader.next()])
	ks_zonn = np.array([int(v) for v in csvreader.next()])		

if to_plot == 1:
	num_samples_to_plot = 1
	colors = 'bg'
	# for n in range(num_samples_to_plot):
	n=3
	latent_draw, = plt.step(times[n,:],cum_vals[n,:],'b-',alpha=.5,where='post',label='Posterior sample')
	print type(latent_draw)
	# plt.plot(ks,N*ps,'r-',linewidth=2)
	pdf_zonn, = plt.plot(ks_zonn,N*ps_zonn,'r-',linewidth=2,label='Zonneveld')
	obs_points, = plt.plot(t,ys[0,:],'ko', label='Observed counts')
	plt.xlim((min(t)-1,max(t)+1))
	plt.ylim((0,20))
	plt.xlabel('Days after May 1st')
	plt.ylabel('Abundance')
	plt.legend(handles=[latent_draw,pdf_zonn,obs_points])
	plt.title('Step function of abundance over time, N = %d' % N)
	plt.savefig('1_draw.pdf',format='pdf')
elif to_plot == 2:

	latent_draw, = plt.plot(0,0,'b-',label='Posterior sample')
	for n in range(cum_vals.shape[0]):
		plt.step(times[n,:],cum_vals[n,:],'b-',alpha=.04,where='post')
	
	# pdf, = plt.plot(ks,N*ps,'r-',linewidth=2,label='pdf')
	pdf_zonn, = plt.plot(ks_zonn,N*ps_zonn,'r-',linewidth=2,label='Zonneveld')
	obs_points, = plt.plot(t,ys[0,:],'ko',fillstyle='none',mew=2, ms=15, label='Observed counts')
	plt.legend(handles=[latent_draw,pdf_zonn,obs_points])
	plt.xlim((min(t)-1,max(t)+1))
	plt.ylim((0,20))
	plt.xlabel('Days after May 1st')
	plt.ylabel('Abundance')
	plt.title('Step function of abundance over time, %d draws, N = %d, ' % (num_samples,N))
	plt.savefig('200_draws.pdf',format='pdf')
elif to_plot == 3:

	t_plot = range(-5,max(t)+1)
	lower = np.zeros(len(t_plot))
	upper = np.zeros(len(t_plot))
	for i,time in enumerate(t_plot):
		argsort = [np.argsort(time - times[s,time-times[s,:]<0]) for s in range(num_samples)]
		for a in range(len(argsort)):
			if argsort[a].shape == (0,):
				argsort[a] = np.array([0])

		sorted_vals = np.sort(cum_vals[range(num_samples),[a[0] for a in argsort]])
		lower[i] = sorted_vals[num_samples*.025]
		upper[i] = sorted_vals[num_samples*.975]

	plt.plot(t_plot,lower,'b-')
	plt.plot(t_plot,upper,'c-')
	plt.plot(ks,N*ps,'r-',linewidth=2)
	plt.plot(t,ys[0,:],'ko')
	plt.xlim((min(t)-1,max(t)+1))
	plt.title('N = %d' % N)

plt.show()

File: old/experiments/latent_variable_experiment.m
function latent_variable_experiment()

no_plot = true;

%parameters learned for population A1 in the Gross et al paper
mu     = 6.9;
sigma  = 3.4;
lambda = 1/0.645;
alpha  = 1;%.32;
N      = 70;
t      = 1:20;
params = struct('N', N, 'alpha', alpha, 't', t);
theta = struct('mu', mu, 'sigma', sigma, 'lambda', lambda);

moves  = {'pairwise','shuffle','mergesplit'};
mcmc_nIter = 1000;
use_ARS = false;
num_runs = 3;

%whether to draw samples from prior or posterior
prior = false;

postfix = '_posterior';
if prior
    postfix = '_prior';
end

if ~no_plot
    figure; hold
end
cum_vals_vec = zeros(num_runs,2*N);
times_vec = zeros(num_runs,2*N);
y_vec = zeros(num_runs,length(t));
for n = 1:num_runs
    
    % get q
    [y, state_0] = sampleState(theta,params);
    
    if prior == true
        q = state_0.q;
    else
        state = mcmc(y,state_0,params,'iterations',mcmc_nIter,'moves',moves,'use_ARS',logical(use_ARS));
        q = state(end).q;
    end
    
    draws = sample_from_q(q,params,theta);
    
    %set up observed curve
    vals = [ones(N,1); -1*ones(N,1)];
    times = [draws(:,1); draws(:,2)];
    [times,inds] = sort(times);
    vals = vals(inds);
    cum_vals = cumsum(vals);
    
    cum_vals_vec(n,:) = cum_vals;
    times_vec(n,:) = times;
    y_vec(n,:) = y;
    
    if ~no_plot
        stairs(times,cum_vals,'edgealpha',.2)
    end
end

%get pdf
ks = floor(min(times)):ceil(max(times));
ps = zeros(size(ks));
for k_ind = 1:length(ks)
    k = ks(k_ind);
    %     func = @(s) 1/(sigma*sqrt(2*pi))*exp(-(s-mu).^2/(2*sigma^2)).*(1 - exp(-1/lambda*(k-s)));
    func = @(s) normpdf(s,mu,sigma).*expcdf(k-s,lambda,'upper');
    ps(k_ind) = integral(func, -inf, k);
end

if ~no_plot
    plot(ks,ps*N,'linewidth',4);
    title(sprintf('N = %d',N))
end

csvwrite([fileparts(which(mfilename)) '/latent_variable_cum_vals',postfix,'.csv'],cum_vals_vec)
csvwrite([fileparts(which(mfilename)) '/latent_variable_times',postfix,'.csv'],times_vec)
csvwrite([fileparts(which(mfilename)) '/latent_variable_ys',postfix,'.csv'],y_vec)
csvwrite([fileparts(which(mfilename)) '/latent_variable_mean_ps_ks',postfix,'.csv'],[ps;ks])



function draws = sample_from_q(q,params,theta)

t = params.t;
N = params.N;
mu = theta.mu;
sigma = theta.sigma;
lambda = theta.lambda;

% how many of each q bin we've filled so far
counts = zeros(size(q));

% list of birth and death times
draws = zeros(N,2); % [birth time, death time]
draw_ind = 1;

while ~all(all(counts == q)) % until we've filled up counts to match q
    
    % sample birth and death time
    birth = normrnd(mu,sigma);
    death = birth + exprnd(lambda);
    
    %determine birth and death intervals
    birth_interval = min(max(floor(birth), 0), max(t));
    death_interval = min(max(floor(death), 0), max(t));
    
    %increment count if its not full
    if counts(birth_interval+1,death_interval+1) < q(birth_interval+1,death_interval+1)
        counts(birth_interval+1,death_interval+1) = counts(birth_interval+1,death_interval+1) + 1;
        
        %add to draws
        draws(draw_ind,:) = [birth,death];
        draw_ind = draw_ind + 1;
    end
    
end
